src/Database/LSMTree/Internal/WriteBufferWriter.hs

```diff
@@ -0,0 +1,227 @@
+import           Database.LSMTree.Internal.Serialise (SerialisedKey,
+                     SerialisedValue)
+import           Database.LSMTree.Internal.WriteBuffer (WriteBuffer)
+import qualified Database.LSMTree.Internal.WriteBuffer as WB
+import           Database.LSMTree.Internal.WriteBufferBlobs (WriteBufferBlobs)
+import qualified Database.LSMTree.Internal.WriteBufferBlobs as WBB
+import qualified System.FS.API as FS
+import           System.FS.API (HasFS)
+import qualified System.FS.BlockIO.API as FS
+import           System.FS.BlockIO.API (HasBlockIO)
+
+
+writeWriteBuffer ::
+     (MonadSTM m, MonadST m, MonadThrow m)
+  => HasFS m h
+  -> HasBlockIO m h
+  -> WriteBufferFsPaths
+  -> WriteBuffer
+  -> WriteBufferBlobs m h
+  -> m ()
+writeWriteBuffer hfs hbio fsPaths buffer blobs = do
+  writer <- new hfs hbio fsPaths
+  for_ (WB.toList buffer) $ \(key, op) ->
+    -- TODO: The fmap entry here reallocates even when there are no blobs.
+    addKeyOp writer key (WBB.mkRawBlobRef blobs <$> op)
+  void $ unsafeFinalise True writer
+
+-- data SerialisedWriteBuffer m h = SerialisedWriteBuffer
+--     { -- | The file system paths for all the files used by the serialised write buffer.
+--       serialisedWriteBufferFsPaths    :: !WriteBufferFsPaths
+--       -- | The (read mode) file handles.
+--     , serialisedWriteBufferHandles    :: !(ForWriteBufferFiles (FS.Handle h))
+--     , serialisedWriteBufferHasFS      :: !(HasFS m h)
+--     , serialisedWriteBufferHasBlockIO :: !(HasBlockIO m h)
+--     }
+
+data WriteBufferWriter m h = WriteBufferWriter
+  { -- | The file system paths for all the files used by the serialised write buffer.
+    writeBufferWriterFsPaths    :: !WriteBufferFsPaths,
+    -- | The page accumulator.
+    writeBufferWriterPageAcc    :: !(PageAcc (PrimState m)),
+    -- | The byte offset within the blob file for the next blob to be written.
+    writeBufferWriterBlobOffset :: !(PrimVar (PrimState m) Word64),
+    -- | The (write mode) file handles.
+    writeBufferWriterHandles    :: !(ForWriteBufferFiles (ChecksumHandle (PrimState m) h)),
+    writeBufferWriterHasFS      :: !(HasFS m h),
+    writeBufferWriterHasBlockIO :: !(HasBlockIO m h)
+  }
+
+{-# SPECIALISE new ::
+     HasFS IO h
+  -> HasBlockIO IO h
+  -> WriteBufferFsPaths
+  -> IO (WriteBufferWriter IO h) #-}
+-- | Create a 'WriteBufferWriter' to start serialising a 'WriteBuffer'.
+new ::
+     (MonadST m, MonadSTM m)
+  => HasFS m h
+  -> HasBlockIO m h
+  -> WriteBufferFsPaths
+  -> m (WriteBufferWriter m h)
+new hfs hbio fsPaths = do
+  writeBufferWriterPageAcc <- ST.stToIO PageAcc.newPageAcc
+  writeBufferWriterBlobOffset <- newPrimVar 0
+  writeBufferWriterHandles <-
+    traverse (makeHandle hfs) (pathsForWriteBufferFiles fsPaths)
+  return WriteBufferWriter
+    { writeBufferWriterFsPaths    = fsPaths,
+      writeBufferWriterHasFS      = hfs,
+      writeBufferWriterHasBlockIO = hbio,
+      ..
+    }
+
+-- | See 'RunBuilder.unsafeFinalise'.
```

Comment:
`RunBuilder` is not imported, so you'll have to use fully qualified names, like `Database.LSMTree.Internal.RunBuilder.unsafeFinalise`